client/apiclient.go

```diff
@@ -47,13 +50,82 @@ type TLSConfig struct {
 }
 
 // NewAPIClient gets an API client for a product.
-func NewAPIClient(product, baseURL string, headers map[string]string) *APIClient {
+func NewAPIClient(product, baseURL string, headers map[string]string, t TLSConfig) (*APIClient, error) {
+	tlsClientConfig, err := createTLSClientConfig(t)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsClientConfig,
+		},
+	}
```

Comment:
the construction of the client and its transport here in NewAPIClient (for now at least) was exactly what I had in mind, glad it makes sense to ya here too :)

the only (vague) concern I have is that the other aspects of `http.DefaultTransport`, which `http.DefaultClient` uses, are excluded here.  I don't know to what extent they are actually important, but I think I would like to retain those defaults.

we can either copy-paste the whole [stdlib default `Transport` struct](https://cs.opensource.google/go/go/+/dev.boringcrypto.go1.17:src/net/http/transport.go;l=43-54) into here, or use it where it is and add to it, e.g.

```go
	transport := http.DefaultTransport.(*tls.Transport)
	transport.TLSClientConfig = tlsClientConfig
	client := &http.Client{
		Transport: transport,
	}
```

fwiw in the past @mkcp and I had agreed that copy-pasting the struct here is nice for explicitness and easier future flexibility if we want to change our defaults.

Reply:
I have pushed a change that I think helps address this concern, ensuring we start from a `DefaultTransport` and attaching a TLS Client Config if needed. Per out-of-band discussion, it looks like the implementation of `DefaultTransport` changes between go 1.17 and 1.18, such that it feels like it could be a little brittle to copy/paste the struct definition into our code. If we find this to be worthwhile in the future, it may be worth re-investigating it then.